### ISSUE TITLE
Serialization impls

### ISSUE THREAD

**Issue by mzabaluev** [dhardy/hashindexed#2]
I'd like to add serialization/deserialization impls for Serde.

With serialized data structures, there's sometimes a mismatch between the storage format that is a sequence of structures, and an efficient in-memory representation which wants the structures to be indexed by some "primary key" intrinsic to the structures. It would be nice to have a sequence deserialize directly into `HashIndexed`.

**Comment by dhardy** [comment]
`HashIndexed` is little more than a wrapper around `HashSet`. Does that implement serialisation?

What kind of API were you thinking of?

```
pub fn serialize(&self, w: &mut Write) -> Result<()> {...}
pub fn deserialize(r: &mut Read) -> Result<HashIndexed<...>> { ... }
```

The latter should possibly also complain if `r` is not empty after finishing, or return the number of bytes read or something.

Also, due to hash randomisation, the same `HashIndexed` could result in many different serialised data streams.

**Comment by mzabaluev** [comment]
Serde and rustc_serialize provide the serialization/deserialization traits for `HashSet`, because it's such a central type. With non-standard libraries, there is no alternative to providing the serialization traits in the same crate that provides the types themselves.

Serde does not work with raw I/O streams, it's polymorphic on various format serializers/deserializers. Check out the documentation for [Serde](https://github.com/serde-rs/serde) on the API required. Once the serialization/deserialization traits are implemented, `IndexedHash` can be used inside data structures so that serialization for enclosing types can be generated automatically by the compiler plugin.

**Comment by dhardy** [comment]
So it only needs a `#derive(Serialize, Deserialize)` line?

Great, send a pull request (if possible, with a test case).

**Comment by mzabaluev** [comment]
> So it only needs a `#derive(Serialize, Deserialize)` line?

No, that would derive a dumb struct-like serialization featuring the `set` field.
Instead, there should be manual implementations delegating to `HashSet`'s.

**Comment by dhardy** [comment]
What's wrong with dumb? It would serialise all data and restore later. Send a PR or at least a good example of what you're trying to achieve.

BTW I already do serialisation of `HashIndexed` data in Pippin — but not by adding the commands to `HashIndexed` since Pippin has its own "serialised data" formats. Depending on the application you might need to do it this way.
